vlab: generate wiring for externals

edipascale opened the pull request:
Generate wirings for externals as part of hhfab vlab gen. Add parameters to determine how many externals and external connections are created, and where the ext. connections are placed (i.e. mclag, eslag or orphan leaves).

All attachments are currently vlan tagged, I didn't want to over complicate things with extra params for untagged.
Also all externals are in the same namespace, and attachments are created for all externals and all connections.

Closes: #521

edipascale commented:
I've added commits to also address #522, but we have a problem here:
- I've set the default to 1 external and 1 connection on an eslag leaf, similar to what we have in physical envs
- however, in collapsed-core mode only mclag switches are created, which means that the defaults are now invalid for this mode
- and on the other hand, we don't want to default to mclag externals because of #524 

In terms of having the ci passing, it would be easy enough to add explicit values to the vlab-gen for either the collapsed-core (if we keep the defaults as described above) or the spine-leaf cases (if we do not touch the current default). The question is what is the "least undesirable" option @Frostman

Frostman commented on pkg/hhfab/vlabrunner.go:
```diff
@@ -613,6 +615,24 @@ func (c *Config) VLABRun(ctx context.Context, vlab *VLAB, opts VLABRunOpts) erro
+					// TODO make it configurable
+					if err := c.SetupPeerings(ctx, vlab, SetupPeeringsOpts{
+						WaitSwitchesReady: true,
+						Requests:          []string{"1+4", "3~external-01:s=subnet-01"},
```
Well, number of vpcs depends on the config as you can specify 0 servers and so no vpcs would be generated.

Frostman replied:
I think the only safe thing would be to just start with the "1~default:s=subnet-01" (assuming first external is renamed to `default`. It's guaranteed that we'll have one VPC. We can improve it later.

edipascale replied:
I'm fine with the change (minus the default naming for reasons explained below), but I assumed the on-ready commands were just a convenience hack to run hard-coded commands in the CI without spawning a new process, in which case this is never going to be called by anyone else... so does it really matter? Wouldn't you have the same problem for the onReady setup-vpcs if you had no servers? We could always hide the command if needed or make it configurable (and pass the appropriate parameters in the CI), but personally I'd just leave it as is

Frostman reviewed (changes requested):
We should probably generated conns to external from the the from both switches of the last redundancy group configured. Let's chat next week and discuss what would be the best approach.

Frostman commented on pkg/hhfab/vlabbuilder.go:
```diff
@@ -614,9 +667,71 @@ func (b *VLABBuilder) Build(ctx context.Context, l *apiutil.Loader, fabricMode m
+	if b.ExtCount > 0 {
+		externals := []vpcapi.External{}
+		extAsn := 64102
+		inboundAsnID := 65102
```
ASN will depend on the specific leaf, but in case of communities those doesn't have to match the ASN, it's just a separate value, would be nice to rename to avoid misunderstanding

Frostman commented on pkg/hhfab/vlabbuilder.go:
```diff
@@ -614,9 +667,71 @@ func (b *VLABBuilder) Build(ctx context.Context, l *apiutil.Loader, fabricMode m
+		communityRuleID := 1000
+
+		for i := uint8(1); i <= b.ExtCount; i++ {
+			externalName := fmt.Sprintf("external-%02d", i)
```
I know it sounds a bit ugly, but let's name the first one `default` so it would work with the `release-test` out of the box which is probably one of the most important goals

edipascale replied:
unfortunately we can't, because I'm using the external name as the name of the VRF in the virtual external VM and `default` is precisely the only reserved VRF name :) once we address the remaining issues I'll change the named external in the release test accordingly

Frostman commented on pkg/hhfab/vlabbuilder.go:
```diff
@@ -728,3 +843,41 @@ func (b *VLABBuilder) createConnection(ctx context.Context, spec wiringapi.Conne
+func (b *VLABBuilder) createExternal(ctx context.Context, name string, spec vpcapi.ExternalSpec) (*vpcapi.External, error) {
+	external := &vpcapi.External{
+		TypeMeta: kmetav1.TypeMeta{
+			Kind:       "External", //?
```
re comment - are you looking for a const for it?

edipascale replied:
exactly, I could not find it in fabric

Frostman commented on cmd/hhfab/main.go:
```diff
@@ -245,6 +246,28 @@ func Run(ctx context.Context) error {
+			Name:        "external-eslag-connections",
+			Usage:       "number of external connections from ESLAG switches",
+			Destination: &wgExtESLAGConns,
+			Value:       1,
```
let's make it 2 by default so we're having redundant link

edipascale commented:
current status:
* the changes in release-test to use the first external have to be reverted or at least adapted; we cannot use virtual externals for most scenarios, as multiple external peerings (together with the limitation on virtual switches and ACLs) mean that VPCs will be able to ping each other even when they shouldn't. We can either revert the change, or fetch the annotations for externals and discard non-hardware ones.
* on top of that, there is https://github.com/githedgehog/internal/issues/145 which makes test-connectivity jobs with the new external peering fail in the ci

edipascale commented:
> current status:
> 
> * the changes in release-test to use the first external have to be reverted or at least adapted; we cannot use virtual externals for most scenarios, as multiple external peerings (together with the limitation on virtual switches and ACLs) mean that VPCs will be able to ping each other even when they shouldn't. We can either revert the change, or fetch the annotations for externals and discard non-hardware ones.

I went for the second option but it's not perfect. Technically the issue is not with virtual externals, but with virtual switches attached to any type of externals. Because right now we only have hardware externals connected to hardware switches and virtual externals connected to virtual switches, discarding virtual externals is the easy solution... but we probably want to revisit it.

> * on top of that, there is [troubleshoot virtual external peering with active connections on multiple leaves internal#145](https://github.com/githedgehog/internal/issues/145) which makes test-connectivity jobs with the new external peering fail in the ci

I've already got an answer from BCM on the case I opened requesting more info, which I provided; let's see if we can get unblocked there. In the meanwhile I reduced the autogenerated external connections from 2 to 1, which should work even with this strange issue.

Also I'm seeing several failures in the CI, both here and in other PRs. but they do not appear related to the changes. I'll dig a bit more tomorrow.

edipascale commented:
@Frostman this is technically ready to review now. I've removed the default external generation as discussed, and added warnings in case an option with multiple external connections is selected. The only thing missing is a run on a physical environment to check whether we need any additional fix for https://github.com/githedgehog/lab/issues/247, unsure whether we want to wait for that before tagging this as ready
